File: src/main/java/com/braydenoneal/blogicmod/block/ModBlocks.java
package com.braydenoneal.blogicmod.block;

import com.braydenoneal.blogicmod.BlogicMod;
import com.braydenoneal.blogicmod.item.ModItemGroup;
import net.fabricmc.fabric.api.item.v1.FabricItemSettings;
import net.fabricmc.fabric.api.object.builder.v1.block.FabricBlockSettings;
import net.minecraft.block.Block;
import net.minecraft.block.BlockState;
import net.minecraft.block.Material;
import net.minecraft.item.BlockItem;
import net.minecraft.item.Item;
import net.minecraft.item.ItemGroup;
import net.minecraft.util.Identifier;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.registry.Registry;
import net.minecraft.world.BlockView;

public class ModBlocks {
	public static final Block DIODE = registerBlock("diode",
			new DiodeBlock(FabricBlockSettings.of(Material.STONE).strength(1f)
					.solidBlock(ModBlocks::never)),
			ModItemGroup.BLOGIC);

	public static final Block LIGHT_EMITTER = registerBlock("light_emitter",
			new LightEmitterBlock(FabricBlockSettings.of(Material.STONE).strength(1f)
					.solidBlock(ModBlocks::never)),
			ModItemGroup.BLOGIC);

	public static final Block TOGGLE = registerBlock("toggle",
			new ToggleBlock(FabricBlockSettings.of(Material.STONE).strength(1f)
					.solidBlock(ModBlocks::never)),
			ModItemGroup.BLOGIC);

	public static final Block AND_GATE = registerBlock("and_gate",
			new AndGateBlock(FabricBlockSettings.of(Material.STONE).strength(1f)
					.solidBlock(ModBlocks::never)),
			ModItemGroup.BLOGIC);

	public static final Block OR_GATE = registerBlock("or_gate",
			new OrGateBlock(FabricBlockSettings.of(Material.STONE).strength(1f)
					.solidBlock(ModBlocks::never)),
			ModItemGroup.BLOGIC);

	public static final Block XOR_GATE = registerBlock("xor_gate",
			new XorGateBlock(FabricBlockSettings.of(Material.STONE).strength(1f)
					.solidBlock(ModBlocks::never)),
			ModItemGroup.BLOGIC);

	public static final Block INVERTER = registerBlock("inverter",
			new InverterBlock(FabricBlockSettings.of(Material.STONE).strength(1f)
					.solidBlock(ModBlocks::never)),
			ModItemGroup.BLOGIC);

	public static final Block CABLE = registerBlock("cable",
			new CableBlock(FabricBlockSettings.of(Material.GLASS).strength(1f)
					.solidBlock(ModBlocks::never)
					.suffocates(ModBlocks::never)
					.blockVision(ModBlocks::never)
					.nonOpaque()
			),
			ModItemGroup.BLOGIC);

	private static Block registerBlock(String name, Block block, ItemGroup itemGroup) {
		registerBlockItem(name, block, itemGroup);
		return Registry.register(Registry.BLOCK, new Identifier(BlogicMod.MOD_ID, name), block);
	}

	private static Item registerBlockItem(String name, Block block, ItemGroup itemGroup) {
		return Registry.register(
				Registry.ITEM, new Identifier(BlogicMod.MOD_ID, name),
				new BlockItem(block, new FabricItemSettings().group(itemGroup)));
	}

	private static boolean never(BlockState state, BlockView world, BlockPos pos) {
		return false;
	}

	public static void registerModBlocks() {}
}

File: src/main/kotlin/block/ModBlocks.kt
package block

import net.fabricmc.fabric.api.itemgroup.v1.ItemGroupEvents
import net.fabricmc.fabric.api.itemgroup.v1.ItemGroupEvents.ModifyEntries
import net.minecraft.core.Registry
import net.minecraft.core.registries.BuiltInRegistries
import net.minecraft.core.registries.Registries
import net.minecraft.resources.Identifier
import net.minecraft.resources.ResourceKey
import net.minecraft.world.item.BlockItem
import net.minecraft.world.item.CreativeModeTabs
import net.minecraft.world.item.Item
import net.minecraft.world.item.Items
import net.minecraft.world.level.block.Block
import net.minecraft.world.level.block.Blocks
import net.minecraft.world.level.block.state.BlockBehaviour
import java.util.function.Function

object ModBlocks {
    val CABLE: Block = register("cable", { CableBlock(it) }, CableBlock.settings())
    val CONTROLLER: Block = register("controller", { ControllerBlock(it) }, BlockBehaviour.Properties.ofFullCopy(Blocks.STONE))

    private fun register(name: String, blockFactory: Function<BlockBehaviour.Properties, Block>, settings: BlockBehaviour.Properties): Block {
        val blockKey = ResourceKey.create(Registries.BLOCK, Identifier.fromNamespaceAndPath(Blogic.MOD_ID, name))
        val block = blockFactory.apply(settings.setId(blockKey))
        val itemKey = ResourceKey.create(Registries.ITEM, Identifier.fromNamespaceAndPath(Blogic.MOD_ID, name))
        val blockItem = BlockItem(block, Item.Properties().setId(itemKey))
        Registry.register(BuiltInRegistries.ITEM, itemKey, blockItem)
        return Registry.register(BuiltInRegistries.BLOCK, blockKey, block)
    }

    fun initialize() {
        ItemGroupEvents.modifyEntriesEvent(CreativeModeTabs.REDSTONE_BLOCKS).register(
            ModifyEntries {
                it.addBefore(Items.REDSTONE, CABLE.asItem())
                it.addBefore(Items.REDSTONE, CONTROLLER.asItem())
            },
        )
    }
}
